### PR TITLE
Implemented .toString() modification

### LINKED ISSUE #2: [BUG] Doesn't change toString() result
Currently, if I run: `rename(function hello() {}, 'stuff')`, it returns:

```js
function hello() {}
```

instead of the expected

```js
function stuff() {}
```

This needs a complete rewrite to the `index.js`. I wrote a possible solution that uses `eval`:

```js
(fn, name) => eval('(() => (' + fn.toString().replace(new RegExp(fn.name), name) + '))()')
```

This seems to give the correct output of:

```js
function stuff() {}
```

The `name` property also gives the expected `'stuff'`. I can create a patch with this change if wanted.

### PR DESCRIPTION
**Pull request by Yash-Singh1**
Fixes #2

Modifies the function's `toString` function.

### CODE REVIEW

**Comment by Yash-Singh1** [comment]
@sindresorhus Looks like `xo` and `ava` don't support NodeJS 8 and underneath. I just tested manually using the REPL and the tests succeeded for NodeJS 4 and above:

```
>nvm install 4
Downloading and installing node v4.9.1...
Downloading https://nodejs.org/dist/v4.9.1/node-v4.9.1-linux-x64.tar.xz...
############################################################################################################################################################################################################ 100.0%
Computing checksum with sha256sum
Checksums matched!
Now using node v4.9.1 (npm v2.15.11)
>node
> const m = require('./index.js')
undefined
> m((hello, hi) => {}, 'nothing').toString()
'(hello, hi) => {}'
> m(function () {}, 'nothing').toString()
'function () {}'
> m(function /*cmon*/ /*sdf*/ man() {}, 'unicorn').toString()
'function unicorn() {}'
> m(function /* cmon */ /* sdf */ man() {}, 'unicorn').toString()
'function unicorn() {}'
> m(function /*cmon*//*sdf*/ man() {}, 'unicorn').toString()
'function unicorn() {}'
> m(function // stuff
... another(){}, 'unicorn').toString()
'function unicorn(){}'
```

Notice how the comments and newlines are removed. For NodeJS 9 and underneath, comments within the function's declaration are emitted from the `.toString()`. That means that running the following:

```js
(function /* hello world */ cool() {}).toString();
```

Will output `'function cool() {}'` in NodeJS 9 while it will output `'function /* hello world */ cool() {}'` in NodeJS 10.

**Comment by Yash-Singh1** [comment]
Taking a look at https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/toString#comparing_actual_source_code_and_tostring_results, there is a lot more work to be done on this PR.

**Comment by Yash-Singh1** [comment]
@sindresorhus As I said above, `xo` and `ava` don't support NodeJS 8 and underneath even though the `package.json` says that `rename-fn` supports NodeJS 4 and above. Because of this, should the linting run only on NodeJS 14 inside GitHub Actions while the testing runs in plain NodeJS instead of `ava` at the actions?

File: index.js [hunk 1]
```diff
@@ -1,2 +1,37 @@
...
+			functionString
+				.replace(/(async|function|\s*)/g, '')
+				.replace(/(\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))/g)
+				.trim()
+				.startsWith('(')
+		) &&
+		functionString.indexOf('class') !== 0 &&
+		(!/^(get|set)/.test(functionString) || !functionString.slice(fn.name.length).replace(/\s/g, '').replace(/(\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*/g, '').trim().startsWith('('))
+	) {
+		const captured = /^((?:get|set)?(?:async\s*)?(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:function)?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\*?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*)([^]*?)\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\(/.exec(functionString);
```
Comment (sindresorhus) [inline index.js]: This regex is way too long and complicated.

File: index.js [hunk 2]
```diff
@@ -1,2 +1,37 @@
+/* eslint-disable func-names, no-extra-bind */
+
 'use strict';
-module.exports = (fn, name) => Object.defineProperty(fn, 'name', {value: name, configurable: true});
+module.exports = (fn, name) => {
+	Object.defineProperty(fn, 'name', {value: name, configurable: true});
+	const functionString = fn.toString();
+	if (
```
Comment (sindresorhus) [inline index.js]: The whole replacement should only be done lazily when `.toString()` is called on the renamed function.

**Comment by sindresorhus** [comment]
> As I said above, xo and ava don't support NodeJS 8 and underneath even though the package.json says that rename-fn supports NodeJS 4 and above. Because of this, should the linting run only on NodeJS 14 inside GitHub Actions while the testing runs in plain NodeJS instead of ava at the actions?

Fixed in https://github.com/sindresorhus/rename-fn/pull/3/commits/bef4357d69c43215dfe1afbfaac295849f526c19. I will properly require Node.js 10 after this PR is merged.

File: index.js (hunk 1 above)
Comment (Yash-Singh1) [reply index.js]: Do you have any suggestions on reducing the size? It is the only way to get the program working.

File: index.js (hunk 2 above)
Comment (Yash-Singh1) [reply index.js]: Do you mean the following?:

```diff
@@ -3,7 +3,6 @@
 'use strict';
 module.exports = (fn, name) => {
 	Object.defineProperty(fn, 'name', {value: name, configurable: true});
-	const functionString = fn.toString();
 	if (
 		!(
 			!/^(?:async\s*)?(?:(\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:function)?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\*?.*?\s*(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n)\s*)*\(/.test(functionString) ||
@@ -16,11 +15,13 @@ module.exports = (fn, name) => {
 		functionString.indexOf('class') !== 0 &&
 		(!/^(get|set)/.test(functionString) || !functionString.slice(fn.name.length).replace(/\s/g, '').replace(/(\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*/g, '').trim().startsWith('('))
 	) {
+		const functionString = fn.toString();
 		const captured = /^((?:get|set)?(?:async\s*)?(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:function)?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\*?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*)([^]*?)\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\(/.exec(functionString);
 		fn.toString = (function toString() {
 			return functionString.slice(0, captured[1].length) + name + functionString.slice(captured[1].length + captured[2].length);
 		}).bind(fn);
 	} else if (functionString.indexOf('class') === 0) {
+		const functionString = fn.toString();
 		const captured = /^(class\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*)([^]*?)\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:extends|{)/.exec(functionString);
 		if (captured[2] !== '') {
 			fn.toString = (function toString() {
@@ -28,6 +29,7 @@ module.exports = (fn, name) => {
 			}).bind(fn);
 		}
 	} else if (/^(get|set)/.test(functionString)) {
+		const functionString = fn.toString();
 		fn.toString = (function toString() {
 			return name + functionString.slice(fn.name.length);
 		}).bind(fn);
```

Comment (sindresorhus) [reply index.js]: No, I mean when the **user** calls `.toString()`.

File: index.js (hunk 1 above)
Comment (sindresorhus) [reply index.js]: Like I commented earlier, strip the comments in an initial pass and then to the replacement.

File: index.js (hunk 2 above)
Comment (Yash-Singh1) [reply index.js]: ae44a71

File: index.js (hunk 1 above)
Comment (Yash-Singh1) [reply index.js]: The current `RegExp`, preserves comments and whitespace. Stripping the comments would not preserve it.

File: index.js [hunk 3]
```diff
@@ -1,2 +1,35 @@
...
+		functionString.indexOf('class') !== 0 &&
+		(!/^(get|set)/.test(functionString) || !functionString.slice(fn.name.length).replace(/\s/g, '').replace(/(\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*/g, '').trim().startsWith('('))
+	) {
+		fn.toString = (function toString() {
+			const captured = /^((?:get|set)?(?:async\s*)?(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:function)?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\*?\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*)([^]*?)\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*\(/.exec(functionString);
+			return functionString.slice(0, captured[1].length) + name + functionString.slice(captured[1].length + captured[2].length);
+		}).bind(fn);
+	} else if (functionString.indexOf('class') === 0 && !functionString.replace('class', '').replace(/\s*/g, '').replace(/\/\*[^]*?\*\/|\/\/[^]*?(\r\n|\r|\n)\s*/g, '').startsWith('{')) {
+		fn.toString = (function toString() {
+			const captured = /^(class\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*)([^]*?)\s*(?:(?:\/\*[^]*?\*\/|\/\/[^]*?(?:\r\n|\r|\n))\s*)*(?:extends|{)/.exec(functionString);
```
Comment (sindresorhus) [inline index.js]: I think the newlines are always `\n`.

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/rename-fn/pull/3/files#r555997325 is still not done. Absolutely nothing toString related should run until the user calls `.toString()` themselves.

**Comment by sindresorhus** [comment]
I would recommend working on simplifying the code and regex. It's not in a mergeable state at the moment.

File: index.js (hunk 3 above)
Comment (Yash-Singh1) [reply index.js]: Yes, the next will probably be covered by the `\s*`

**Comment by Yash-Singh1** [comment]
@sindresorhus I made several commits overall reducing the main function `RegExp` size by 74 characters and the class `RegExp` size by 36 characters and simplifying some other `RegExp`'s by removing non-capturing groups. There was also one edge case that I found and fixed, methods with names function*/async*. Does it look good or do you have any potential improvements?

**Comment by sindresorhus** [comment]
> Absolutely nothing toString related should run until the user calls .toString() themselves.

Still not done.

**Comment by Yash-Singh1** [comment]
@sindresorhus Currently the code structure is like this:

```js
if (condition) {
  fn.toString = () => {
    // ...
  }
}
```

From what I am assuming that you are saying, we want a structure like this:

```js
fn.toString = () => {
  if (condition) {
    // ...
  }
}
```

The reason that I am using the original structure is because the `condition` checks whether the `fn.toString` function even has to be changed. As suggested by the comments I left in the code, the `condition` makes sure that:

The function is:

- Not an arrow function
- Not a function without a name
- Not a class without a name

If one of these three scenarios fail, the `fn.toString` function doesn't even have to be touched.

File: index.js (hunk 1 above)
Comment (Yash-Singh1) [reply index.js]: https://github.com/sindresorhus/rename-fn/pull/3#issuecomment-760578829

File: index.js (hunk 2 above)
Comment (Yash-Singh1) [reply index.js]: Do you mean https://github.com/sindresorhus/rename-fn/pull/3#issuecomment-762419071?

**Comment by sindresorhus** [comment]
It's about optimizing for the common case. Almost no one is going to call `.toString()`, so I don't think everyone should have to pay for the overhead of the check whether `.toString` needs to be replaced.

**Comment by Yash-Singh1** [comment]
22d22b8

**Comment by Yash-Singh1** [comment]
Updated PR and ready for review

**Comment by Yash-Singh1** [comment]
Friendly review bump :)

**Comment by Yash-Singh1** [comment]
Just added more test cases. Another reminder for review.

**Comment by Yash-Singh1** [comment]
@sindresorhus

**Comment by sindresorhus** [comment]
I haven't forgotten about this. I'm just nervous about the regexes. I have had to deal with a lot of [ReDoS](https://en.wikipedia.org/wiki/ReDoS) recently. There's just so many ways to get regexes wrong. I'm leaning towards waiting with this until the `/l` flag can be used: https://v8.dev/blog/non-backtracking-regexp

Maybe you could publish your own fork in the meantime if you need this yourself?

**Comment by sindresorhus** [comment]
I keep getting more ReDoS vulnerabilities for even super simple regex, so there's no way I want to risk this unfortunately. We can pick this up at some point when `/l` exists.

**Pull request closed without merging**